src/RunDebugCodeLens.ts

```diff
@@ -0,0 +1,37 @@
+import * as vscode from 'vscode';
+
+
+export class RunDebugCodeLens implements vscode.CodeLensProvider {
```

Comment:
Cool idea. Just curious, could we provide the code lens directly via LSP (using [`textDocument/codeLens`](https://microsoft.github.io/language-server-protocol/specification#textDocument_codeLens))? Not entirely sure about that though, since the commands for launching a debug session are implemented on the client-side anyway (LSP should be able to handle arbitrary commands, I think, but then we'd need to be careful with clients that don't support this command)

Reply:
I would feel weird about putting it into the language server for a few reasons:
1. Separation of concerns. A language server provides a lot of functionality, but does not have a concept of running or debugging the program (at least in my experience). Those parts are the responsibility of a debug-adapter or other programs. The language server is more of a program to analyze source code and provide helpers for it (completion, linting, symbol information etc.)
2. How would we turn off the code lens where the client does not have a debug adapter enabled? We would have to have controls for each code lens. That feels more like a client responsibility. If we have more code lenses in the future, that will seem cumbersome to me. (I think a code lens for the number of references or similar that the Java plugins have would be cool in kotlin-language-server in the future). 


So in my view, it makes most sense to do it like this 🙂  I don't think the problem here is what LSP can do, but what it SHOULD do. Like you say, if we do it in the language server, it would be a minor hazzle to handle for clients that don't support it. (a minor annoyance at worst probably, but still)

Reply:
That sounds totally reasonable, thanks for the explanation!